### DISCUSSION TITLE
Allow Standard Value Types to be Specified and Used as Domain Concepts

### DISCUSSION THREAD

**Discussion by kabua** [dotnet/csharplang#6681] [Language Ideas]
Ever since .NET 2.0, I have been thinking about this language feature when I realized that Value Types like `char`, `int`, `long`,  `double`, etc all define 1) the memory footprint and 2) the range of values these types can take on. Then we switch gears and create higher-level concepts using things like: `struct` and `class`, and we give these constructs names that come from the application domain (ignoring for the moment that these also define where the memory will be allocated). 

Take, for example, you are writing an e-commerce website. Then some of your domain concepts might include things like: customer, address, zip code, shopping cart, age, quantity, price, etc. For things like: customer, address, and shopping cart, you could model these domain concepts as classes, as in; `public class Customer` and  `public class ShoppingCart` etc. But what about things like: zip code, age, quantity, price, etc.?

Well, today, we would define these using memory storage concepts like `string`, `int`, and `double`. The problem with these names is that they all come from the compiler and hardware domain and not from the business domain. Therefore, I have always felt that the standard Value Types have caused an impedance mismatch between the business domain and the hardware domain.

For almost any given method signature, we constantly co-mingle these two domains. For example:
```
class Address
{
    Address(string city, string state)
}
```
In this very simple `ctor` example, we switch back and forth from the business domain to the hardware domain 5 times. 

A few days ago, we updated some of our older code to the latest version of our internal framework, which had a few breaking changes. Therefore our older code had to change from something like this:
```
void OpenDatabase(string databaseIpAddress, string databaseName) { ... }
```
to
```
void Use(string databaseName) { ... }
```

unfortunately, the code was upgraded from 
```
myobject.OpenDatabase(Env.SalesDatabaseAddress, Constants.DatabaseNames.Sales)
````
to
```
configBuilder.Use(Env.SalesDatabaseAddress)
```

Can you spot the bug? This simple mistake cost the company 16 developer hours (4 devs times four hours) to find this simple mistake.

This is just one of many examples which happen almost every day. Remember when [NASA crashed the Mars Climate Orbiter](http://www.cnn.com/TECH/space/9909/30/mars.metric.02/), which cost $125 million due to a metric-imperial impedance mismatch?

### How can We Avoid these Domain Mismatches in the Future?

In C++, we can use `#define`, and in Python, we can use  `TypeVar`, but there isn't anything like this for C#.

The new `record` concept looked promising at first, but it will not work for the standard Value Types as these types would have to be wrapped in a structure, which would lead to loss of functionality for the wrapped Value Type, and it would make accessing the Value Type slower. 

We need something that the compiler can translate directly back to the underlining Value Type, as the C++ does for `#define`, but it needs to be done in a type-safe way. After all, we are talking about C#.

Using the new `record` keyword and concept as inspiration, what if we used `domain` in the same way? Then zip code, age, quantity, price, etc. could be written as true domain concepts backed by explicit Value Type. Thus creating a mapping between the business domain and the hardware domain.

Like so:
```
public domain ZipCode : string;
public domain Age : uint;
public domain quantity : uint;
public domain Price : double;
```
Note: the new `domain` keyword would not only work for the standard Value Types but for other types like DateTime, which are .NET framework types or types from other vendors etc. However, these domain types cannot be extended using interfaces, generics, etc., since the underlining Value Type can't be extended this way either. Except for implicit and explicit typecasting. 

Using the previous example, the signature would then be:
```
void OpenDatabase(DatabaseAddress address, DatabaseName name) { ... }
```
And the new breaking change signature would have been:
```
void Use(DatabaseName name) { ... }
```
And this code would never have compiled:
```
configBuilder.Use(Env.SalesDatabaseAddress)
```
Since a `DatabaseAddress` isn't a `DatabaseName`.

Had NASA used domain concepts instead of `double` perhaps the crash would not have happened. Or what about a life support system where some methods take in two `double`, and the developer swaps the two values around by mistake - this simple mistake could lead to loss of life?

### Typecasting
After prototyping this concept, we noticed that the biggest issue comes from type casing. Casting from a domain object to the underlining type isn't the problem. But casting from a Value Type to a domain object is, and creating an implicit conversion would not help as the compiler would not know which domain object to cast too. Thus the developer would be required to cast a Value Type into the correct domain concept explicitly.

However, developers should be able to control what types can automatically be cast between domain objects. Like so:
```
public domain DateOfBirth : DateTime;
public domain Age : int
{
   public static explicit operator Age(DateOfBirth other) { ... }
}
```

We understand that other considerations must be considered and possibly expand on this concept. Still, we think the initial concept of `domain` would greatly benefit the C# communitee and business.

**Comment by HaloFour** [comment]
See: #1695

IMO you get the majority of this today with record structs:

```cs
public record struct DateOfBirth(DateTime Value);
public record struct Age(int Value) {
    public static explicit operator Age(DateOfBirth dob) { ... }
}
```

**Reply by HaloFour** [reply]
Also, see roles here: #5498

**Reply by kabua** [reply]
@HaloFour I like that others are also thinking along these same lines. This suggests there is a real need here. The problem, as I see it and addressed above, is that single-value records would require a de-referencing of the record, as in: `record.Value` of some sort, and I feel that it would make it feel clunky, awkward, and unnatural to use. I'm also concerned about duplication of code and using RAM inefficiency, so code-gen-ing a new type from an existing type seems to be an overkill. 

Therefore, the proposal of `domain` is very different from `record` in that `domain` doesn't create anything; it IS the underlining type during runtime, just like what the compiler does with types like `System.Integer`, which is compiled to a native 32-bit integer vs. creating a wrapper class around the native 32-bit integer which is what Java uses to do (or perhaps still does).

This way, a dictionary of `Dictionary<string, string>` could be explicitly defined as `Dictionary<IpAddress, MachineName>`, but the underlining type the ASM would see would still be `Dictionary<string, string>` and not two custom string implementations.

The idea of roles is interesting; I will have to think about this more.

**Reply by HaloFour** [reply]
> The problem, as I see it and addressed above, is that single-value records would require a de-referencing of the record, as in: `record.Value` of some sort, and I feel that it would make it feel clunky, awkward, and unnatural to use.

Conversion operators could be used to smooth that dereference, and those could be generated via source generators.

> I'm also concerned about duplication of code and using RAM inefficiency, so code-gen-ing a new type from an existing type seems to be an overkill.

The runtime elides the wrapping struct in these cases making them just as efficient as using the underlying type directly.

> it IS the underlining type during runtime, just like what the compiler does with types like `System.Integer`, which is compiled to a native 32-bit integer vs. creating a wrapper class around the native 32-bit integer which is what Java uses to do (or perhaps still does).

This is because `int` is a C# alias for `System.Int32`, which is just a normal struct.  There are no wrapper types for primitives in thr runtime because .NET doesn't differentiate between them like Java does today.  But this is why you can use `int` and `Int32` interchangeably in code, they are the same thing as far as thr compiler is concerned, so there is no type safety.

This would also be a problem with your proposal, eliding the wrapper type makes it the same as the underlying type as far as the runtime is concerned which would eliminate the type safety this proposal seeks to offer.

**Reply by kabua** [reply]
> The runtime elides the wrapping struct in these cases making them just as efficient as using the underlying type directly.
> 
> > it IS the underlining type during runtime, just like what the compiler does with types like `System.Integer`, which is compiled to a native 32-bit integer vs. creating a wrapper class around the native 32-bit integer which is what Java uses to do (or perhaps still does).
> 
> This is because `int` is a C# alias for `System.Int32`, which is just a normal struct. There are no wrapper types for primitives in thr runtime because .NET doesn't differentiate between them like Java does today. But this is why you can use `int` and `Int32` interchangeably in code, they are the same thing as far as thr compiler is concerned, so there is no type safety.
> 
> This would also be a problem with your proposal, eliding the wrapper type makes it the same as the underlying type as far as the runtime is concerned which would eliminate the type safety this proposal seeks to offer.

Thanks, I understand this, and perhaps you stated it better than I did. In the end, we are saying the same thing.

**Comment by orthoxerox** [comment]
There are at least three slightly different features that I want to see represented:

1. Units of measure. I don't want any additional restrictions on the range of values, but this decimal is the price in minor units, that decimal is the quantity. There's never a reason why I would want to assign one to the other or compare them or pass them to a function that expects two values of generic type T. However, I want to perform arithmetic on them and I want it to make sense.
2. New types as quasi units of measure. I don't want any additional restrictions on the range of values, but this int is a CustomerID, that int is an AccountID and I don't want to mix them up. There's never a reason why I would want to assign one to the other or compare them or pass them to a function that expects two values of generic type T (but I might want to have a function that works on all ints that are SomethingID and only them). I don't even want to perform arithmetic on them and I want the compiler to call me out when I do.
3. New types as types with explicitly restricted range. DeployMode is a string that can be either "cluster" or "client", not "foo", not "cluster ", not "CLUSTER", not null or anything else other than these two values. If I get it from somewhere, I want to be sure it's a valid value.

This discussion looks like it's mostly about 2 with a hint of 1. Record structs, as suggested by @HaloFour, will work for 2, but need a good dose of syntactic frosting for existing libraries to find them palatable (so one can get them from and save them to the database, bind them to UI controls without additional ceremony)

**Reply by kabua** [reply]
@orthoxerox I like how you succinctly stated these 3 points. My original goal was more around points 1 & 2. As in:

1. the application had concepts of Price and Quantity, as you stated. Therefore, in code, these two domain concepts could be specified as:
```
public domain Price : double;
public domain Quantity : double;

void AddToCart(string decription, Price price, Quantity quantity);
```
Thus, the developer can't pass quantity for `Price` or price for `Quantity` as these two parameters are separate and independent types. Without any additional restrictions on their ranges. 

The other benefit is when requirements change. For example, management now states the application can no longer use `double` for things like Price or Quantity and must use `decimal`s. Therefore, the code could be quickly changed to
```
public domain Price : decimal;
public domain Quantity : decimal;
```
and retested for performance. Only to discover that using `decimal` everywhere is very expensive. On further investigation, it is pointed out that performance could be improved (and better scope of the value's range) if we change Quantity from `decimal` to `uint` as "I want 1.25 computers" doesn't make any sense. Thus quantity should be a whole number between Zero and Unit.MAX. Thus in one location, the code could be updated and retested within minutes.

2. I agree CustomerID and AccountID, while both may use `int` as their underlying hardware types, are two unrelated domain concepts. And could be implemented as:
```
public domain CustomerID : int;
public domain AccountID : int;

void ThrowIfLessThanOne(int number);

ThrowIfLessThanOne(CustomerID);          // Compiler error, Type "CustomerID" can not be implicitly converted to type "int".
ThrowIfLessThanOne((int) CustomerID);    // No Compiler error since the developer's intent is very clear.
```

Now, with C# 11, how would we handle this case?
```
void ThrowIfLessThanOne(T number) where INumber<T>;
```
Should `ThrowIfLessThanOne(CustomerID);` throw an error or not? I can see where both cases could be valid. Perhaps a compiler attribute could indicate which option is valid, like we have with other compiler attribute guards. 

3. I never thought of a case like this. Perhaps because we have already created a `StringEnum` type that mimics the functionality of an `enum` but for strings. However, I do agree with you that there are several cases where a `StringEnum` type concept would be extremely useful, and I would like to see this concept be part of the language.

You also bring up another great point, what if we did need to constrain a value to a range? Could we do something like this, then?
```
public domain Quantity : uint [0..^10];
```
Where `Quantity` would be constrained to the values zero to `unit.max - 10`.  Extending this example to the next step, we could get:
```
public domain DeployMode : string ["cluster", "client"];
```
While the syntax appears easy enough to add, the complexity of the compiler just jumped X-fold. Therefore, depending on the adoption rate of the new `domain` keyword and concept, these extended features could be added later without affecting the current design.

**Comment by JiriZidek** [comment]
Why not just copy/paste F# functionality ?

**Reply by HaloFour** [reply]
You mean single-case unions?  C# already has very similar syntax for this:

F#:
```fsharp
type EmailAddress = EmailAddress of string
```

C#:
```cs
public record struct EmailAddress(string Value);
```

F# does have a bit more functionality, but a lot of that is driven through extremely custom metadata and hard references on F#'s core assemblies.  It couldn't be copy-pastaed, but a lot of it is being considered for discriminated unions.

**Reply by JiriZidek** [reply]
I meant more "design", not directly copying src ! And it may be a good "V2" avoiding pitfalls in F#.
